Source/WTF/wtf/RangeAdaptors.h

```diff
@@ -0,0 +1,68 @@
+/*
+ *  Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Library General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Library General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Library General Public License
+ *  along with this library; see the file COPYING.LIB.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301, USA.
+ *
+ */
+
+#pragma once
+
+#include <ranges>
+
+namespace WTF {
+
+struct FromRange {
+    explicit FromRange() = default;
+};
+
+// Global instance for use as tag
+inline constexpr FromRange fromRange { };
+
+template <typename Derived>
+struct RangeAdaptorClosure {
+    template <std::ranges::input_range Range>
+    friend auto operator|(Range&& range, const Derived& adaptor)
+    {
+        return adaptor(std::forward<Range>(range));
+    }
+
+    template <typename OtherClosure>
+    friend auto operator|(const Derived& left, const OtherClosure& right)
+    {
+        return [=](auto&& range) {
+            return right(left(std::forward<decltype(range)>(range)));
+        };
+    }
+};
+
+template <typename Container>
+struct FromRangeConverter : RangeAdaptorClosure<FromRangeConverter<Container>> {
+    template <std::ranges::input_range Range>
+        requires requires(Range&& range) { Container(fromRange, std::forward<Range>(range)); }
```

Comment:
aside: something i've noticed that global objects `fromRange` has a storage cost. not sure if i'm mimicking the right setup here:

https://godbolt.org/z/8fMeTdWxb

```
fromRange:
        .zero   1
```

rather, if you throw a temporary, that cost is gone

```cpp
foo.add<int>(FromRange { }, a, b);
```

perhaps, all of this doesn't matter when compiled with optimization flags.